src/ReportManager/DetailsSection/SchemaForm/index.js

```diff
@@ -15,14 +15,38 @@ export const FIELDS = {
   [FORM_ELEMENT_TYPES.TEXT]: Text,
 };
 
-const SchemaForm = ({ formData, onFormChange, onFormSubmit, renderSubmitButton, schema }) => {
-  const fields = useMemo(() => schema ? makeFieldsFromSchema(schema) : {}, [schema]);
+const SchemaForm = ({ formData, onFormDataChange, onFormSubmit, renderSubmitButton, schema }) => {
+  const runValidations = useSchemaValidations(schema);
+
+  const [fieldErrors, setFieldErrors] = useState({});
+
+  const fields = useMemo(() => makeFieldsFromSchema(schema), [schema]);
+
+  const fieldValues = useMemo(() => Object.entries(formData).reduce((accumulator, [fieldId, fieldValue]) => {
+    // TODO: Collections will require recusivity here.
+
+    return { ...accumulator, [fieldId]: fieldValue };
+  }, {}), [formData]);
+
+  const onFieldChange = useCallback((fieldId, value) => {
+    // TODO: Collections will require recusivity here.
+    // If the value is empty, set it as undefined so AJV validation returns errors for required fields.
+    const newFormData = { ...formData, [fieldId]: value || undefined };
+
+    onFormDataChange(newFormData);
+    setFieldErrors((fieldErrors) => ({ ...fieldErrors, [fieldId]: undefined }));
+  }, [formData, onFormDataChange]);
 
   const onSubmit = (event) => {
     event.preventDefault();
 
-    // TODO: Add form validation with AJV before submission and show errors.
-    onFormSubmit({ formData });
+    const fieldErrors = runValidations(formData);
+    if (fieldErrors) {
+      setFieldErrors(fieldErrors);
+      document.getElementById(Object.keys(fieldErrors)[0]).focus();
```

Comment:
This auto focus feels kind of non-react stuff, using a regular ref was complicating things?

Reply:
Can you explain a bit more what do you mean by non-react? About using refs, we would need a ref for every field, and we don't know how many fields there are. That is dynamic depending on the schema. Maybe I can refactor it to be a little bit more legible:
```
const idOfFirstFieldWithError = Object.keys(fieldErrors)[0];
document.getElementById(idOfFirstFieldWithError).focus();
```
I don't know if that helps.

Reply:
I meant non-react by accessing directly to a node using `document.getElementById` When usually We do it with refs, and as you mentioned it, having refs it would cause having unnecessary stuff around all components, so yes, I agree with the little refactor 🫡

Reply:
Got you. Yeah it's not common to get elements by id in React. But in this specific scenario is pretty handy since we can't have a fixed number of refs because the fields are dynamic. I'll send the changes soon!